// @recogito/annotorious/packages/annotorious-react/dist/annotorious-react.es.js
import { Annotorious as P, AnnotoriousContext as C, useAnnotationStore as y, useAnnotations as v, useAnnotator as w, useAnnotatorUser as B, useSelection as I, useViewportState as O } from "./annotorious-react.es2.js";
import { AnnotoriousPlugin as $ } from "./annotorious-react.es3.js";
import { Draggable as F } from "./annotorious-react.es4.js";
import { ImageAnnotator as W } from "./annotorious-react.es5.js";
import "./annotorious-react.es6.js";
import "./annotorious-react.es7.js";
import { defaultColorProvider as o } from "./annotorious-react.es8.js";
import "./annotorious-react.es9.js";
import { PointerSelectAction as e } from "./annotorious-react.es10.js";
import { createBody as r } from "./annotorious-react.es11.js";
import { ShapeType as t, W3CImageFormat as n } from "./annotorious-react.es12.js";
import { OpenSeadragonAnnotator as D, useViewer as H } from "./annotorious-react.es13.js";
import { OpenSeadragonPopup as j } from "./annotorious-react.es14.js";
import { OpenSeadragonViewer as q } from "./annotorious-react.es15.js";
const f = o, d = e, l = r, A = t, S = n;
export {
  P as Annotorious,
  C as AnnotoriousContext,
  $ as AnnotoriousPlugin,
  F as Draggable,
  W as ImageAnnotator,
  D as OpenSeadragonAnnotator,
  j as OpenSeadragonPopup,
  q as OpenSeadragonViewer,
  d as PointerSelectAction,
  A as ShapeType,
  S as W3CImageFormat,
  l as createBody,
  f as defaultColorProvider,
  y as useAnnotationStore,
  v as useAnnotations,
  w as useAnnotator,
  B as useAnnotatorUser,
  I as useSelection,
  H as useViewer,
  O as useViewportState
};
