// @recogito/annotorious/packages/annotorious-svelte/dist/annotorious-svelte.es3.js
import { SvelteComponent as g, init as b, safe_not_equal as h, empty as k, insert as w, transition_in as l, transition_out as c, check_outros as A, detach as O, create_slot as v, update_slot_base as C, get_all_dirty_from_scope as q, get_slot_changes as y, group_outros as D, setContext as m } from "./annotorious-svelte.es5.js";
import { toSvelteStore as N } from "./annotorious-svelte.es6.js";
import "./annotorious-svelte.es7.js";
import { createOSDAnnotator as T } from "./annotorious-svelte.es8.js";
function d(s) {
  let n;
  const o = (
    /*#slots*/
    s[4].default
  ), t = v(
    o,
    s,
    /*$$scope*/
    s[3],
    null
  );
  return {
    c() {
      t && t.c();
    },
    m(e, i) {
      t && t.m(e, i), n = !0;
    },
    p(e, i) {
      t && t.p && (!n || i & /*$$scope*/
      8) && C(
        t,
        o,
        e,
        /*$$scope*/
        e[3],
        n ? y(
          o,
          /*$$scope*/
          e[3],
          i,
          null
        ) : q(
          /*$$scope*/
          e[3]
        ),
        null
      );
    },
    i(e) {
      n || (l(t, e), n = !0);
    },
    o(e) {
      c(t, e), n = !1;
    },
    d(e) {
      t && t.d(e);
    }
  };
}
function j(s) {
  let n, o, t = (
    /*viewer*/
    s[0] && d(s)
  );
  return {
    c() {
      t && t.c(), n = k();
    },
    m(e, i) {
      t && t.m(e, i), w(e, n, i), o = !0;
    },
    p(e, [i]) {
      /*viewer*/
      e[0] ? t ? (t.p(e, i), i & /*viewer*/
      1 && l(t, 1)) : (t = d(e), t.c(), l(t, 1), t.m(n.parentNode, n)) : t && (D(), c(t, 1, 1, () => {
        t = null;
      }), A());
    },
    i(e) {
      o || (l(t), o = !0);
    },
    o(e) {
      c(t), o = !1;
    },
    d(e) {
      t && t.d(e), e && O(n);
    }
  };
}
function z(s, n, o) {
  let { $$slots: t = {}, $$scope: e } = n, { viewer: i } = n, { opts: f = {} } = n, { anno: a = void 0 } = n;
  const p = (r) => {
    if (r) {
      const u = T(r, f), S = N(u.state.store), _ = {
        ...u,
        state: { ...u.state, store: S }
      };
      m("anno", _), m("viewer", r), o(1, a = _);
    }
  };
  return s.$$set = (r) => {
    "viewer" in r && o(0, i = r.viewer), "opts" in r && o(2, f = r.opts), "anno" in r && o(1, a = r.anno), "$$scope" in r && o(3, e = r.$$scope);
  }, s.$$.update = () => {
    s.$$.dirty & /*viewer*/
    1 && p(i);
  }, [i, a, f, e, t];
}
class B extends g {
  constructor(n) {
    super(), b(this, n, z, j, h, { viewer: 0, opts: 2, anno: 1 });
  }
}
const I = B;
export {
  I as default
};
