// @recogito/annotorious/packages/annotorious-svelte/dist/annotorious-svelte.es2.js
import { SvelteComponent as O, init as T, safe_not_equal as q, empty as w, insert as E, transition_in as u, transition_out as _, check_outros as A, detach as L, getContext as N, component_subscribe as S, onMount as X, create_slot as Y, element as j, attr as c, update_slot_base as z, get_all_dirty_from_scope as B, get_slot_changes as D, group_outros as F } from "./annotorious-svelte.es5.js";
const G = (o) => ({ hovered: o & /*hovered*/
16 }), b = (o) => ({ hovered: (
  /*hovered*/
  o[4]
) });
function g(o) {
  let r, s, e;
  const t = (
    /*#slots*/
    o[8].default
  ), n = Y(
    t,
    o,
    /*$$scope*/
    o[7],
    b
  );
  return {
    c() {
      r = j("div"), n && n.c(), c(r, "class", "a9s-tooltip"), c(r, "style", s = `left:${/*left*/
      o[2]}px; top:${/*top*/
      o[1]}px; position: absolute;`);
    },
    m(i, l) {
      E(i, r, l), n && n.m(r, null), e = !0;
    },
    p(i, l) {
      n && n.p && (!e || l & /*$$scope, hovered*/
      144) && z(
        n,
        t,
        i,
        /*$$scope*/
        i[7],
        e ? D(
          t,
          /*$$scope*/
          i[7],
          l,
          G
        ) : B(
          /*$$scope*/
          i[7]
        ),
        b
      ), (!e || l & /*left, top*/
      6 && s !== (s = `left:${/*left*/
      i[2]}px; top:${/*top*/
      i[1]}px; position: absolute;`)) && c(r, "style", s);
    },
    i(i) {
      e || (u(n, i), e = !0);
    },
    o(i) {
      _(n, i), e = !1;
    },
    d(i) {
      i && L(r), n && n.d(i);
    }
  };
}
function H(o) {
  let r, s, e = (
    /*$hover*/
    o[0] && /*show*/
    o[3] && g(o)
  );
  return {
    c() {
      e && e.c(), r = w();
    },
    m(t, n) {
      e && e.m(t, n), E(t, r, n), s = !0;
    },
    p(t, [n]) {
      /*$hover*/
      t[0] && /*show*/
      t[3] ? e ? (e.p(t, n), n & /*$hover, show*/
      9 && u(e, 1)) : (e = g(t), e.c(), u(e, 1), e.m(r.parentNode, r)) : e && (F(), _(e, 1, 1, () => {
        e = null;
      }), A());
    },
    i(t) {
      s || (u(e), s = !0);
    },
    o(t) {
      _(e), s = !1;
    },
    d(t) {
      e && e.d(t), t && L(r);
    }
  };
}
function I(o, r, s) {
  let e, t, { $$slots: n = {}, $$scope: i } = r, { container: l } = r;
  const k = N("anno"), { store: M, hover: p } = k.state;
  S(o, p, (f) => s(0, t = f));
  let v, d, a = !0;
  return X(() => {
    const f = () => {
      s(3, a = !0);
    }, m = (y) => {
      const { offsetX: P, offsetY: C } = y;
      s(2, d = P), s(1, v = C);
    }, h = () => {
      s(3, a = !1);
    };
    return l.addEventListener("pointerenter", f), l.addEventListener("pointermove", m), l.addEventListener("pointerleave", h), () => {
      l.removeEventListener("pointerenter", f), l.removeEventListener("pointermove", m), l.removeEventListener("pointerleave", h);
    };
  }), o.$$set = (f) => {
    "container" in f && s(6, l = f.container), "$$scope" in f && s(7, i = f.$$scope);
  }, o.$$.update = () => {
    o.$$.dirty & /*$hover*/
    1 && s(4, e = t ? M.getAnnotation(t) : void 0);
  }, [t, v, d, a, e, p, l, i, n];
}
class J extends O {
  constructor(r) {
    super(), T(this, r, I, H, q, { container: 6 });
  }
}
const Q = J;
export {
  Q as default
};
