// @iiif/vault/dist/actions/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var typesafeActions = require('typesafe-actions');

const IMPORT_ENTITIES = "@iiif/IMPORT_ENTITIES";
const MODIFY_ENTITY_FIELD = "@iiif/MODIFY_ENTITY_FIELD";
const REORDER_ENTITY_FIELD = "@iiif/REORDER_ENTITY_FIELD";
const ADD_REFERENCE = "@iiif/ADD_REFERENCE";
const REMOVE_REFERENCE = "@iiif/REMOVE_REFERENCE";
const importEntities = typesafeActions.createAction(IMPORT_ENTITIES)();
const modifyEntityField = typesafeActions.createAction(MODIFY_ENTITY_FIELD)();
const reorderEntityField = typesafeActions.createAction(REORDER_ENTITY_FIELD)();
const addReference = typesafeActions.createAction(ADD_REFERENCE)();
const removeReference = typesafeActions.createAction(REMOVE_REFERENCE)();
const entityActions = { importEntities, modifyEntityField, reorderEntityField, addReference, removeReference };

const ADD_MAPPING = "@iiif/ADD_MAPPING";
const ADD_MAPPINGS = "@iiif/ADD_MAPPINGS";
const addMapping = typesafeActions.createAction(ADD_MAPPING)();
const addMappings = typesafeActions.createAction(ADD_MAPPINGS)();
const mappingActions = { addMapping, addMappings };

const SET_META_VALUE = "@iiif/SET_META_VALUE";
const SET_META_VALUE_DYNAMIC = "@iiif/SET_META_VALUE_DYNAMIC";
const UNSET_META_VALUE = "@iiif/UNSET_META_VALUE";
const setMetaValue = typesafeActions.createAction(SET_META_VALUE)();
const setMetaValueDynamic = typesafeActions.createAction(SET_META_VALUE_DYNAMIC)();
const unsetMetaValue = typesafeActions.createAction(UNSET_META_VALUE)();
const metaActions = {
  setMetaValue,
  setMetaValueDynamic,
  unsetMetaValue
};

const RESOURCE_ERROR = "RESOURCE_ERROR";
const RESOURCE_LOADING = "RESOURCE_LOADING";
const RESOURCE_READY = "RESOURCE_READY";
const REQUEST_RESOURCE = "@iiif/REQUEST_RESOURCE";
const REQUEST_ERROR = "@iiif/REQUEST_ERROR";
const REQUEST_MISMATCH = "@iiif/REQUEST_MISMATCH";
const REQUEST_COMPLETE = "@iiif/REQUEST_COMPLETE";
const REQUEST_OFFLINE_RESOURCE = "@iiif/REQUEST_OFFLINE_RESOURCE";
const requestResource = typesafeActions.createAction(REQUEST_RESOURCE)();
const requestError = typesafeActions.createAction(REQUEST_ERROR)();
const requestMismatch = typesafeActions.createAction(REQUEST_MISMATCH)();
const requestComplete = typesafeActions.createAction(REQUEST_COMPLETE)();
const requestOfflineResource = typesafeActions.createAction(REQUEST_OFFLINE_RESOURCE)();
const requestActions = {
  requestResource,
  requestError,
  requestMismatch,
  requestComplete,
  requestOfflineResource
};

const BATCH_ACTIONS = "@iiif/BATCH";
const BATCH_IMPORT = "@iiif/BATCH_IMPORT";
const batchActions = typesafeActions.createAction(BATCH_ACTIONS)();
const batchImport = typesafeActions.createAction(BATCH_IMPORT)();

exports.ADD_MAPPING = ADD_MAPPING;
exports.ADD_MAPPINGS = ADD_MAPPINGS;
exports.ADD_REFERENCE = ADD_REFERENCE;
exports.BATCH_ACTIONS = BATCH_ACTIONS;
exports.BATCH_IMPORT = BATCH_IMPORT;
exports.IMPORT_ENTITIES = IMPORT_ENTITIES;
exports.MODIFY_ENTITY_FIELD = MODIFY_ENTITY_FIELD;
exports.REMOVE_REFERENCE = REMOVE_REFERENCE;
exports.REORDER_ENTITY_FIELD = REORDER_ENTITY_FIELD;
exports.REQUEST_COMPLETE = REQUEST_COMPLETE;
exports.REQUEST_ERROR = REQUEST_ERROR;
exports.REQUEST_MISMATCH = REQUEST_MISMATCH;
exports.REQUEST_OFFLINE_RESOURCE = REQUEST_OFFLINE_RESOURCE;
exports.REQUEST_RESOURCE = REQUEST_RESOURCE;
exports.RESOURCE_ERROR = RESOURCE_ERROR;
exports.RESOURCE_LOADING = RESOURCE_LOADING;
exports.RESOURCE_READY = RESOURCE_READY;
exports.SET_META_VALUE = SET_META_VALUE;
exports.SET_META_VALUE_DYNAMIC = SET_META_VALUE_DYNAMIC;
exports.UNSET_META_VALUE = UNSET_META_VALUE;
exports.addMapping = addMapping;
exports.addMappings = addMappings;
exports.addReference = addReference;
exports.batchActions = batchActions;
exports.batchImport = batchImport;
exports.entityActions = entityActions;
exports.importEntities = importEntities;
exports.mappingActions = mappingActions;
exports.metaActions = metaActions;
exports.modifyEntityField = modifyEntityField;
exports.removeReference = removeReference;
exports.reorderEntityField = reorderEntityField;
exports.requestActions = requestActions;
exports.requestComplete = requestComplete;
exports.requestError = requestError;
exports.requestMismatch = requestMismatch;
exports.requestOfflineResource = requestOfflineResource;
exports.requestResource = requestResource;
