// cloverIIIF/dist/scroll/index.mjs
import r, { createContext as q, useReducer as ie, useContext as S, useState as I, useRef as B, useEffect as v } from "react";
import { Vault as L } from "@iiif/vault";
import ae from "sanitize-html";
import { createStitches as se } from "@stitches/react";
import "hls.js";
import K from "openseadragon";
import "@iiif/vault-helpers";
import { ErrorBoundary as le } from "react-error-boundary";
import { v4 as Y } from "uuid";
import ce from "@nulib/use-markdown";
import de from "flexsearch";
const A = {
  annotations: [],
  manifest: void 0,
  searchString: void 0,
  options: {
    offset: 0
  },
  vault: new L()
};
function ue(e, t) {
  switch (t.type) {
    case "updateAnnotations":
      return {
        ...e,
        annotations: t.payload
      };
    case "updateSearchString":
      return {
        ...e,
        searchString: t.payload
      };
    default:
      return e;
  }
}
const w = q({
  dispatch: () => null,
  state: A
}), me = (e) => {
  const { children: t, manifest: n } = e, a = {
    ...A.options,
    ...e.options
  }, [o, i] = ie(ue, A);
  return /* @__PURE__ */ r.createElement(
    w.Provider,
    {
      value: {
        state: { ...o, manifest: n, options: a },
        dispatch: i
      }
    },
    t
  );
}, pe = (e, t = "none") => {
  if (!e)
    return null;
  if (typeof e == "string")
    return [e];
  if (!e[t]) {
    const n = Object.getOwnPropertyNames(e);
    if (n.length > 0)
      return e[n[0]];
  }
  return !e[t] || !Array.isArray(e[t]) ? null : e[t];
}, M = (e, t = "none", n = ", ") => {
  const a = pe(e, t);
  return Array.isArray(a) ? a.join(`${n}`) : a;
};
function he(e) {
  return { __html: ge(e) };
}
function T(e, t) {
  const n = Object.keys(e).filter(
    (o) => t.includes(o) ? null : o
  ), a = new Object();
  return n.forEach((o) => {
    a[o] = e[o];
  }), a;
}
function ge(e) {
  return ae(e, {
    allowedAttributes: {
      a: ["href"],
      img: ["alt", "src", "height", "width"]
    },
    allowedSchemes: ["http", "https", "mailto"],
    allowedTags: [
      "a",
      "b",
      "br",
      "i",
      "img",
      "p",
      "small",
      "span",
      "sub",
      "sup"
    ]
  });
}
const k = 209, fe = {
  colors: {
    /*
     * Black and dark grays in a light theme.
     * Must contrast to 4.5 or greater with `secondary`.
     */
    primary: "#1a1d1e",
    primaryMuted: "#26292b",
    primaryAlt: "#151718",
    /*
     * Key brand color(s).
     * Must contrast to 4.5 or greater with `secondary`.
     */
    accent: `hsl(${k} 100% 38.2%)`,
    accentMuted: `hsl(${k} 80% 61.8%)`,
    accentAlt: `hsl(${k} 80% 30%)`,
    /*
     * White and light grays in a light theme.
     * Must contrast to 4.5 or greater with `primary` and  `accent`.
     */
    secondary: "#FFFFFF",
    secondaryMuted: "#e6e8eb",
    secondaryAlt: "#c1c8cd"
  },
  fontSizes: {
    1: "12px",
    2: "13px",
    3: "15px",
    4: "17px",
    5: "19px",
    6: "21px",
    7: "27px",
    8: "35px",
    9: "59px"
  },
  lineHeights: {
    1: "12px",
    2: "13px",
    3: "15px",
    4: "17px",
    5: "19px",
    6: "21px",
    7: "27px",
    8: "35px",
    9: "59px"
  },
  sizes: {
    1: "5px",
    2: "10px",
    3: "15px",
    4: "20px",
    5: "25px",
    6: "35px",
    7: "45px",
    8: "65px",
    9: "80px"
  },
  space: {
    1: "5px",
    2: "10px",
    3: "15px",
    4: "20px",
    5: "25px",
    6: "35px",
    7: "45px",
    8: "65px",
    9: "80px"
  },
  radii: {
    1: "4px",
    2: "6px",
    3: "8px",
    4: "12px",
    round: "50%",
    pill: "9999px"
  },
  transitions: {
    all: "all 300ms cubic-bezier(0.16, 1, 0.3, 1)"
  },
  zIndices: {
    1: "100",
    2: "200",
    3: "300",
    4: "400",
    max: "999"
  }
}, xe = {
  xxs: "(max-width: 349px)",
  xs: "(max-width: 575px)",
  sm: "(max-width: 767px)",
  md: "(max-width: 991px)",
  lg: "(max-width: 90rem)",
  xl: "(min-width: calc(90rem + 1px))"
}, { styled: d, css: Xt, keyframes: Gt, createTheme: qt } = se({
  theme: fe,
  media: xe
}), ve = d("span", {}), F = (e) => {
  const { as: t, label: n } = e, o = T(e, ["as", "label"]);
  return /* @__PURE__ */ r.createElement(ve, { as: t, ...o }, M(n, o.lang));
};
d("img", { objectFit: "cover" });
d("a", {});
const we = {
  delimiter: ", "
}, R = q(void 0), ye = () => {
  const e = S(R);
  if (e === void 0)
    throw new Error(
      "usePrimitivesContext must be used with a PrimitivesProvider"
    );
  return e;
}, Se = ({
  children: e,
  initialState: t = we
}) => {
  const n = Ee(t, "delimiter");
  return /* @__PURE__ */ r.createElement(R.Provider, { value: { delimiter: n } }, e);
}, Ee = (e, t) => Object.hasOwn(e, t) ? e[t].toString() : void 0, be = d("span", {}), V = (e) => {
  const { as: t, markup: n } = e, { delimiter: a } = ye();
  if (!n)
    return /* @__PURE__ */ r.createElement(r.Fragment, null);
  const i = T(e, ["as", "markup"]), s = he(
    M(n, i.lang, a)
  );
  return /* @__PURE__ */ r.createElement(be, { as: t, ...i, dangerouslySetInnerHTML: s });
}, Ie = (e) => r.useContext(R) ? /* @__PURE__ */ r.createElement(V, { ...e }) : /* @__PURE__ */ r.createElement(Se, null, /* @__PURE__ */ r.createElement(V, { ...e }));
d("span", {});
d("dl", {});
d("li", {});
d("ul", {});
d("li", {});
d("ul", {});
d("dl", {});
d("li", {});
d("ul", {});
const Ce = (e) => {
  const { as: t, summary: n } = e, o = T(e, ["as", "customValueDelimiter", "summary"]);
  return /* @__PURE__ */ r.createElement(Ie, { as: t, markup: n, ...o });
}, _ = d("div", {
  display: "flex",
  flexDirection: "column",
  justifyContent: "space-between"
}), ke = d("aside", {
  margin: "0",
  padding: "0",
  position: "relative",
  zIndex: 2,
  flexGrow: "1",
  flexShrink: "0",
  [`& ${_}`]: {
    position: "absolute",
    width: "50%",
    top: 0
  },
  "&.anchor": {
    [`& ${_}`]: {
      position: "fixed",
      width: "50%"
    }
  }
}), $e = d("header", {
  display: "flex",
  justifyContent: "space-between",
  fontSize: "1",
  flexGrow: "1",
  flexShrink: "0",
  marginBottom: "1.618rem",
  strong: {
    fontWeight: "400",
    fontSize: "1.33rem"
  }
}), ze = d("section", {
  margin: "0",
  gap: "1rem"
}), Le = ({ label: e }) => /* @__PURE__ */ r.createElement($e, null, /* @__PURE__ */ r.createElement("strong", null, /* @__PURE__ */ r.createElement(F, { label: e }))), Ae = d("article", {
  transition: "all 0.382s ease-in-out",
  display: "flex",
  flexDirection: "row",
  flexWrap: "nowrap",
  gap: "2.618rem"
}), J = d("div", {
  transition: "$all",
  width: "50%",
  opacity: 0,
  transform: "translateX(2.618rem)",
  zIndex: -1
}), _e = d("div", {
  width: "50%",
  display: "flex",
  flexDirection: "column",
  justifyContent: "flex-start"
}), Pe = d("hr", {
  margin: "0",
  borderColor: "transparent",
  height: "1.618rem",
  position: "relative",
  zIndex: 0,
  display: "flex",
  justifyContent: "flex-end",
  marginTop: "2.618rem",
  "&::before": {
    content: "attr(aria-label)",
    position: "absolute",
    right: "1.618rem",
    bottom: "0",
    zIndex: 1,
    display: "flex",
    fontSize: "0.7222rem",
    fontWeight: "400",
    lineHeight: "1rem",
    background: "inherit",
    opacity: 0.7
  },
  "&::after": {
    content: "",
    width: "100%",
    position: "absolute",
    zIndex: 0,
    height: "1px",
    background: "#6662"
  }
}), Be = d("div", {
  position: "relative",
  zIndex: "1",
  display: "flex",
  flexDirection: "column",
  gap: "2.618rem",
  "&[data-figures-visible='true']": {
    [`& ${J}`]: {
      opacity: 1,
      zIndex: 0,
      transform: "translateX(0)"
    }
  }
}), Me = ({
  canvas: e,
  canvasInfo: { current: t, total: n }
}) => /* @__PURE__ */ r.createElement("figcaption", null, /* @__PURE__ */ r.createElement("em", null, t, " / ", n), (e == null ? void 0 : e.label) && /* @__PURE__ */ r.createElement(F, { label: e == null ? void 0 : e.label }), (e == null ? void 0 : e.summary) && /* @__PURE__ */ r.createElement(Ce, { summary: e == null ? void 0 : e.summary, as: "p" }));
let y = window.OpenSeadragon;
if (!y && (y = K, !y))
  throw new Error("OpenSeadragon is missing.");
const W = "http://www.w3.org/2000/svg";
y.Viewer && (y.Viewer.prototype.svgOverlay = function() {
  return this._svgOverlayInfo ? this._svgOverlayInfo : (this._svgOverlayInfo = new Q(this), this._svgOverlayInfo);
});
const Q = function(e) {
  const t = this;
  this._viewer = e, this._containerWidth = 0, this._containerHeight = 0, this._svg = document.createElementNS(W, "svg"), this._svg.style.position = "absolute", this._svg.style.left = 0, this._svg.style.top = 0, this._svg.style.width = "100%", this._svg.style.height = "100%", this._viewer.canvas.appendChild(this._svg), this._node = document.createElementNS(W, "g"), this._svg.appendChild(this._node), this._viewer.addHandler("animation", function() {
    t.resize();
  }), this._viewer.addHandler("open", function() {
    t.resize();
  }), this._viewer.addHandler("rotate", function() {
    t.resize();
  }), this._viewer.addHandler("flip", function() {
    t.resize();
  }), this._viewer.addHandler("resize", function() {
    t.resize();
  }), this.resize();
};
Q.prototype = {
  // ----------
  node: function() {
    return this._node;
  },
  // ----------
  resize: function() {
    this._containerWidth !== this._viewer.container.clientWidth && (this._containerWidth = this._viewer.container.clientWidth, this._svg.setAttribute("width", this._containerWidth)), this._containerHeight !== this._viewer.container.clientHeight && (this._containerHeight = this._viewer.container.clientHeight, this._svg.setAttribute("height", this._containerHeight));
    const e = this._viewer.viewport.pixelFromPoint(new y.Point(0, 0), !0), t = this._viewer.viewport.getZoom(!0), n = this._viewer.viewport.getRotation(), a = this._viewer.viewport.getFlip(), o = this._viewer.viewport._containerInnerSize.x;
    let i = o * t;
    const s = i;
    a && (i = -i, e.x = -e.x + o), this._node.setAttribute(
      "transform",
      "translate(" + e.x + "," + e.y + ") scale(" + i + "," + s + ") rotate(" + n + ")"
    );
  },
  // ----------
  onClick: function(e, t) {
    new y.MouseTracker({
      element: e,
      clickHandler: t
    }).setTracking(!0);
  }
};
const Te = (e) => fetch(`${e.replace(/\/$/, "")}/info.json`).then((t) => t.json()).then((t) => t).catch((t) => {
  console.error(
    `The IIIF tilesource ${e.replace(
      /\/$/,
      ""
    )}/info.json failed to load: ${t}`
  );
}), Fe = (e) => {
  let t, n;
  if (Array.isArray(e) && (t = e[0], t)) {
    let a;
    "@id" in t ? a = t["@id"] : a = t.id, n = a;
  }
  return n;
};
var C = /* @__PURE__ */ ((e) => (e.TiledImage = "tiledImage", e.SimpleImage = "simpleImage", e))(C || {});
const Re = (e) => {
  const t = Array.isArray(e == null ? void 0 : e.service) && (e == null ? void 0 : e.service.length) > 0, n = t ? Fe(e == null ? void 0 : e.service) : e == null ? void 0 : e.id, a = t ? C.TiledImage : C.SimpleImage;
  return {
    uri: n,
    imageType: a
  };
}, Oe = (e, t) => {
  const n = t ? C.TiledImage : C.SimpleImage;
  return {
    uri: e,
    imageType: n
  };
}, He = d("div", {
  display: "flex",
  flexDirection: "column",
  alignItems: "center"
}), Ve = d("p", {
  fontWeight: "bold",
  fontSize: "x-large"
}), We = d("span", {
  fontSize: "medium"
}), De = ({ error: e }) => {
  const { message: t } = e;
  return /* @__PURE__ */ r.createElement(He, { role: "alert" }, /* @__PURE__ */ r.createElement(Ve, { "data-testid": "headline" }, "Something went wrong"), t && /* @__PURE__ */ r.createElement(We, null, `Error message: ${t}`, " "));
}, P = d("div", {
  position: "absolute !important",
  zIndex: "1",
  top: "1rem",
  left: "1rem",
  width: "161.8px",
  height: "100px",
  backgroundColor: "#000D",
  boxShadow: "5px 5px 5px #0002",
  borderRadius: "3px",
  ".displayregion": {
    border: " 3px solid $accent !important",
    boxShadow: "0 0 3px #0006"
  },
  "@sm": {
    width: "123px",
    height: "76px"
  },
  "@xs": {
    width: "100px",
    height: "61.8px"
  }
}), je = d("div", {
  position: "relative",
  width: "100%",
  height: "100%",
  zIndex: "0"
}), Ne = d("div", {
  width: "100%",
  height: "100%",
  maxHeight: "100vh",
  background: "transparent",
  backgroundSize: "contain",
  color: "white",
  position: "relative",
  zIndex: "0",
  overflow: "hidden",
  variants: {
    hasNavigator: {
      true: {
        [`${P}`]: {
          display: "block"
        }
      },
      false: {
        [`${P}`]: {
          display: "none"
        }
      }
    }
  }
}), Ze = d("button", {
  display: "flex",
  height: "2rem",
  width: "2rem",
  borderRadius: "2rem",
  padding: "0",
  margin: "0",
  fontFamily: "inherit",
  background: "none",
  backgroundColor: "$primary",
  border: "none",
  color: "white",
  cursor: "pointer",
  marginLeft: "0.618rem",
  filter: "drop-shadow(2px 2px 5px #0003)",
  transition: "$all",
  boxSizing: "content-box !important",
  "&:first-child": {
    marginLeft: "0"
  },
  "@xs": {
    marginBottom: "0.618rem",
    marginLeft: "0",
    "&:last-child": {
      marginBottom: "0"
    }
  },
  svg: {
    height: "60%",
    width: "60%",
    padding: "20%",
    fill: "$secondary",
    stroke: "$secondary",
    filter: "drop-shadow(2px 2px 5px #0003)",
    transition: "$all",
    boxSizing: "inherit"
  },
  "&:hover, &:focus": {
    backgroundColor: "$accent"
  },
  "&[data-button=rotate-right]": {
    "&:hover, &:focus": {
      svg: {
        rotate: "45deg"
      }
    }
  },
  "&[data-button=rotate-left]": {
    transform: "scaleX(-1)",
    "&:hover, &:focus": {
      svg: {
        rotate: "45deg"
      }
    }
  },
  "&[data-button=reset]": {
    "&:hover, &:focus": {
      svg: {
        rotate: "-15deg"
      }
    }
  }
}), b = ({ className: e, id: t, label: n, children: a }) => {
  const o = n.toLowerCase().replace(/\s/g, "-");
  return /* @__PURE__ */ r.createElement(
    Ze,
    {
      id: t,
      className: e,
      "data-testid": "openseadragon-button",
      "data-button": o
    },
    /* @__PURE__ */ r.createElement(
      "svg",
      {
        xmlns: "http://www.w3.org/2000/svg",
        "aria-labelledby": `${t}-svg-title`,
        "data-testid": "openseadragon-button-svg",
        focusable: "false",
        viewBox: "0 0 512 512",
        role: "img"
      },
      /* @__PURE__ */ r.createElement("title", { id: `${t}-svg-title` }, n),
      a
    )
  );
}, Ue = d("div", {
  position: "absolute",
  zIndex: "1",
  top: "1rem",
  right: "1rem",
  display: "flex",
  "@xs": {
    flexDirection: "column",
    zIndex: "2"
  },
  variants: {
    hasPlaceholder: {
      true: {
        right: "3.618rem",
        "@xs": {
          top: "3.618rem",
          right: "1rem"
        }
      },
      false: {
        right: "1rem",
        "@xs": {
          top: "1rem",
          right: "1rem"
        }
      }
    }
  }
}), Xe = {
  behavior: "smooth",
  block: "center"
}, f = {
  annotationOverlays: {
    backgroundColor: "#6666ff",
    borderColor: "#000099",
    borderType: "solid",
    borderWidth: "1px",
    opacity: "0.5",
    renderOverlays: !0,
    zoomLevel: 2
  },
  background: "transparent",
  canvasBackgroundColor: "#6662",
  canvasHeight: "500px",
  contentSearch: {
    searchResultsLimit: 20,
    overlays: {
      backgroundColor: "#ff6666",
      borderColor: "#990000",
      borderType: "solid",
      borderWidth: "1px",
      opacity: "0.5",
      renderOverlays: !0,
      zoomLevel: 4
    }
  },
  ignoreCaptionLabels: [],
  pages: {
    show: !0,
    toggleLabel: "Pages"
  },
  informationPanel: {
    vtt: {
      autoScroll: {
        enabled: !0,
        settings: Xe
      }
    },
    open: !0,
    renderAbout: !0,
    renderSupplementing: !0,
    renderToggle: !0,
    renderAnnotation: !0,
    renderContentSearch: !0
  },
  openSeadragon: {},
  requestHeaders: { "Content-Type": "application/json" },
  showDownload: !0,
  showIIIFBadge: !0,
  showTitle: !0,
  withCredentials: !1,
  localeText: {
    contentSearch: {
      tabLabel: "Search Results",
      formPlaceholder: "Enter search words",
      noSearchResults: "No search results",
      loading: "Loading...",
      moreResults: "more results"
    }
  }
};
function Ge(e) {
  let t = {
    ...f.informationPanel.vtt.autoScroll
  };
  return typeof e == "object" && (t = "enabled" in e ? e : { enabled: !0, settings: e }), e === !1 && (t.enabled = !1), qe(t.settings), t;
}
function qe({ behavior: e, block: t }) {
  const n = ["auto", "instant", "smooth"], a = ["center", "end", "nearest", "start"];
  if (!n.includes(e))
    throw TypeError(`'${e}' not in ${n.join(" | ")}`);
  if (!a.includes(t))
    throw TypeError(`'${t}' not in ${a.join(" | ")}`);
}
var Z, U;
const Ke = Ge(
  (U = (Z = f == null ? void 0 : f.informationPanel) == null ? void 0 : Z.vtt) == null ? void 0 : U.autoScroll
);
var X, G;
const ee = {
  activeCanvas: "",
  activeManifest: "",
  OSDImageLoaded: !1,
  collection: {},
  configOptions: f,
  customDisplays: [],
  plugins: [],
  isAutoScrollEnabled: Ke.enabled,
  isAutoScrolling: !1,
  isInformationOpen: (X = f == null ? void 0 : f.informationPanel) == null ? void 0 : X.open,
  showPageNavigation: (G = f == null ? void 0 : f.pages) == null ? void 0 : G.show,
  isLoaded: !1,
  isUserScrolling: void 0,
  vault: new L(),
  contentSearchVault: new L(),
  openSeadragonViewer: null,
  viewerId: Y()
}, Ye = r.createContext(ee), Je = r.createContext(ee);
function D() {
  const e = r.useContext(Ye);
  if (e === void 0)
    throw new Error("useViewerState must be used within a ViewerProvider");
  return e;
}
function te() {
  const e = r.useContext(Je);
  if (e === void 0)
    throw new Error("useViewerDispatch must be used within a ViewerProvider");
  return e;
}
const Qe = () => /* @__PURE__ */ r.createElement(
  "path",
  {
    strokeLinecap: "round",
    strokeMiterlimit: "10",
    strokeWidth: "45",
    d: "M256 112v288M400 256H112"
  }
), et = () => /* @__PURE__ */ r.createElement(
  "path",
  {
    strokeLinecap: "round",
    strokeMiterlimit: "10",
    strokeWidth: "45",
    d: "M400 256H112"
  }
), tt = () => /* @__PURE__ */ r.createElement(
  "path",
  {
    fill: "none",
    stroke: "currentColor",
    strokeLinecap: "round",
    strokeLinejoin: "round",
    strokeWidth: "32",
    d: "M432 320v112H320M421.8 421.77L304 304M80 192V80h112M90.2 90.23L208 208M320 80h112v112M421.77 90.2L304 208M192 432H80V320M90.23 421.8L208 304"
  }
), rt = () => /* @__PURE__ */ r.createElement("path", { d: "M448 440a16 16 0 01-12.61-6.15c-22.86-29.27-44.07-51.86-73.32-67C335 352.88 301 345.59 256 344.23V424a16 16 0 01-27 11.57l-176-168a16 16 0 010-23.14l176-168A16 16 0 01256 88v80.36c74.14 3.41 129.38 30.91 164.35 81.87C449.32 292.44 464 350.9 464 424a16 16 0 01-16 16z" }), j = () => /* @__PURE__ */ r.createElement(r.Fragment, null, /* @__PURE__ */ r.createElement(
  "path",
  {
    fill: "none",
    strokeLinecap: "round",
    strokeMiterlimit: "10",
    strokeWidth: "45",
    d: "M400 148l-21.12-24.57A191.43 191.43 0 00240 64C134 64 48 150 48 256s86 192 192 192a192.09 192.09 0 00181.07-128"
  }
), /* @__PURE__ */ r.createElement("path", { d: "M464 97.42V208a16 16 0 01-16 16H337.42c-14.26 0-21.4-17.23-11.32-27.31L436.69 86.1C446.77 76 464 83.16 464 97.42z" })), nt = ({
  _cloverViewerHasPlaceholder: e,
  config: t
}) => {
  const n = D(), {
    activeCanvas: a,
    configOptions: o,
    openSeadragonViewer: i,
    plugins: s,
    vault: u,
    activeManifest: l
  } = n, p = u.get({
    id: a,
    type: "Canvas"
  });
  function m() {
    return s.filter((c) => {
      var g;
      return (g = c.imageViewer) == null ? void 0 : g.menu;
    }).map((c, g) => {
      var x, E, O, H;
      const h = (E = (x = c.imageViewer) == null ? void 0 : x.menu) == null ? void 0 : E.component;
      return /* @__PURE__ */ r.createElement(
        h,
        {
          key: g,
          ...(H = (O = c == null ? void 0 : c.imageViewer) == null ? void 0 : O.menu) == null ? void 0 : H.componentProps,
          activeManifest: l,
          canvas: p,
          viewerConfigOptions: o,
          openSeadragonViewer: i,
          useViewerDispatch: te,
          useViewerState: D
        }
      );
    });
  }
  return /* @__PURE__ */ r.createElement(
    Ue,
    {
      "data-testid": "clover-iiif-image-openseadragon-controls",
      hasPlaceholder: e
    },
    t.showZoomControl && /* @__PURE__ */ r.createElement(r.Fragment, null, /* @__PURE__ */ r.createElement(b, { id: t.zoomInButton, label: "zoom in" }, /* @__PURE__ */ r.createElement(Qe, null)), /* @__PURE__ */ r.createElement(b, { id: t.zoomOutButton, label: "zoom out" }, /* @__PURE__ */ r.createElement(et, null))),
    t.showFullPageControl && /* @__PURE__ */ r.createElement(b, { id: t.fullPageButton, label: "full page" }, /* @__PURE__ */ r.createElement(tt, null)),
    t.showRotationControl && /* @__PURE__ */ r.createElement(r.Fragment, null, /* @__PURE__ */ r.createElement(b, { id: t.rotateRightButton, label: "rotate right" }, /* @__PURE__ */ r.createElement(j, null)), /* @__PURE__ */ r.createElement(b, { id: t.rotateLeftButton, label: "rotate left" }, /* @__PURE__ */ r.createElement(j, null))),
    t.showHomeControl && /* @__PURE__ */ r.createElement(b, { id: t.homeButton, label: "reset" }, /* @__PURE__ */ r.createElement(rt, null)),
    m()
  );
}, ot = ({
  ariaLabel: e,
  config: t,
  uri: n,
  _cloverViewerHasPlaceholder: a,
  imageType: o,
  openSeadragonCallback: i
}) => {
  const [s, u] = I(), [l, p] = I(), m = te(), c = B(!1);
  return v(() => (c.current || (c.current = !0, l || p(K(t))), () => l == null ? void 0 : l.destroy()), []), v(() => {
    l && i && i(l);
  }, [l, i]), v(() => {
    l && n !== s && (l == null || l.forceRedraw(), u(n));
  }, [l, s, n]), v(() => {
    if (s && l)
      switch (o) {
        case "simpleImage":
          l == null || l.addSimpleImage({
            url: s
          });
          break;
        case "tiledImage":
          Te(s).then((g) => {
            try {
              if (!g)
                throw new Error(`No tile source found for ${s}`);
              l == null || l.addTiledImage({
                tileSource: g,
                success: () => {
                  typeof m == "function" && m({
                    type: "updateOSDImageLoaded",
                    OSDImageLoaded: !0
                  });
                }
              });
            } catch (h) {
              console.error(h);
            }
          });
          break;
        default:
          l == null || l.close(), console.warn(
            `Unable to render ${s} in OpenSeadragon as type: "${o}"`
          );
          break;
      }
  }, [o, s]), /* @__PURE__ */ r.createElement(
    Ne,
    {
      className: "clover-iiif-image-openseadragon",
      "data-testid": "clover-iiif-image-openseadragon",
      "data-openseadragon-instance": t.id,
      hasNavigator: t.showNavigator
    },
    /* @__PURE__ */ r.createElement(
      nt,
      {
        _cloverViewerHasPlaceholder: a,
        config: t
      }
    ),
    t.showNavigator && /* @__PURE__ */ r.createElement(
      P,
      {
        id: t.navigatorId,
        "data-testid": "clover-iiif-image-openseadragon-navigator"
      }
    ),
    /* @__PURE__ */ r.createElement(
      je,
      {
        id: t.id,
        "data-testid": "clover-iiif-image-openseadragon-viewport",
        role: "img",
        ...e && { "aria-label": e }
      }
    )
  );
};
function it(e) {
  return {
    id: `openseadragon-${e}`,
    navigatorId: `openseadragon-navigator-${e}`,
    loadTilesWithAjax: !0,
    fullPageButton: `fullPage-${e}`,
    homeButton: `reset-${e}`,
    rotateLeftButton: `rotateLeft-${e}`,
    rotateRightButton: `rotateRight-${e}`,
    zoomInButton: `zoomIn-${e}`,
    zoomOutButton: `zoomOut-${e}`,
    showNavigator: !0,
    showFullPageControl: !0,
    showHomeControl: !0,
    showRotationControl: !0,
    showZoomControl: !0,
    navigatorBorderColor: "transparent",
    gestureSettingsMouse: {
      clickToZoom: !0,
      dblClickToZoom: !0,
      pinchToZoom: !0,
      scrollToZoom: !1
    }
  };
}
const at = ({
  _cloverViewerHasPlaceholder: e = !1,
  body: t,
  instanceId: n,
  isTiledImage: a = !1,
  label: o,
  src: i = "",
  openSeadragonCallback: s,
  openSeadragonConfig: u = {}
}) => {
  const l = n || Y(), p = typeof o == "string" ? o : M(o), m = {
    ...it(l),
    ...u
  }, { imageType: c, uri: g } = t ? Re(t) : Oe(i, a);
  return g ? /* @__PURE__ */ r.createElement(le, { FallbackComponent: De }, /* @__PURE__ */ r.createElement(
    ot,
    {
      _cloverViewerHasPlaceholder: e,
      ariaLabel: p,
      config: m,
      imageType: c,
      key: l,
      uri: g,
      openSeadragonCallback: s
    }
  )) : null;
}, st = d("div", {
  width: "100%",
  height: "400px",
  background: "#6662",
  backgroundSize: "contain",
  color: "white",
  position: "relative",
  zIndex: "1",
  overflow: "hidden"
}), lt = ({ body: e, label: t }) => /* @__PURE__ */ r.createElement(st, { "data-testid": "scroll-figure-image" }, /* @__PURE__ */ r.createElement(
  at,
  {
    body: e,
    openSeadragonConfig: { showNavigator: !1, showHomeControl: !1 },
    ...t && { label: t }
  }
)), ct = d("figure", {
  figcaption: {
    display: "flex",
    flexDirection: "column",
    margin: "1.618rem 0 0",
    opacity: 0.9,
    em: {
      fontSize: "0.9em",
      fontStyle: "normal",
      opacity: 0.7
    }
  }
}), dt = (e, t, n, a) => {
  var s, u;
  const o = {
    canvas: void 0,
    accompanyingCanvas: void 0,
    annotationPage: void 0,
    annotations: []
  }, i = (l) => {
    if (l) {
      if (!l.body || !l.motivation) {
        console.error(
          "Invalid annotation after Hyperion parsing: missing either 'body' or 'motivation'",
          l
        );
        return;
      }
      let p = l.body;
      Array.isArray(p) && (p = p[0]);
      const m = e.get(p.id);
      if (!m)
        return;
      switch (n) {
        case "painting":
          return l.target === t.id && l.motivation && l.motivation[0] === "painting" && a.includes(m.type) && (l.body = m), !!l;
        case "supplementing":
          return;
        default:
          throw new Error("Invalid annotation motivation.");
      }
    }
  };
  if (o.canvas = e.get(t), o.canvas && (o.annotationPage = e.get(o.canvas.items[0]), o.accompanyingCanvas = (s = o.canvas) != null && s.accompanyingCanvas ? e.get((u = o.canvas) == null ? void 0 : u.accompanyingCanvas) : void 0), o.annotationPage) {
    const l = e.get(o.annotationPage.items).map((m) => ({
      body: e.get(m.body[0].id),
      motivation: m.motivation,
      type: "Annotation"
    })), p = [];
    l.forEach((m) => {
      m.body.type === "Choice" ? m.body.items.forEach(
        (c) => p.push({
          ...m,
          id: c.id,
          body: e.get(c.id)
        })
      ) : p.push(m);
    }), o.annotations = p.filter(i);
  }
  return o;
}, ut = (e, t) => {
  const n = dt(
    e,
    { id: t, type: "Canvas" },
    "painting",
    ["Image", "Sound", "Video"]
  );
  if (n.annotations.length !== 0 && n.annotations && n.annotations)
    return n.annotations.map(
      (a) => a == null ? void 0 : a.body
    );
}, mt = ({ canvas: e, canvasInfo: t }) => {
  const { state: n } = S(w), { vault: a } = n, o = ut(a, e.id);
  return o ? /* @__PURE__ */ r.createElement(ct, null, o == null ? void 0 : o.map((i) => /* @__PURE__ */ r.createElement(lt, { body: i, key: i == null ? void 0 : i.id, label: e == null ? void 0 : e.label })), /* @__PURE__ */ r.createElement(Me, { canvas: e, canvasInfo: t })) : null;
}, $ = "255, 197, 32", pt = d("div", {
  ul: {
    padding: "1rem"
  },
  li: {
    listStyleType: "disc",
    li: {
      listStyleType: "circle"
    }
  },
  "span.highlight": {
    position: "relative",
    fontWeight: "bold",
    "&::before": {
      top: "0",
      position: "absolute",
      display: "inline",
      content: "",
      width: "calc(100% + 4px)",
      height: "calc(100% + 2px) ",
      marginLeft: "-2px",
      borderRadius: "3px",
      border: `1px solid rgba(${$}, 0.2)`,
      borderBottom: `1px solid rgba(${$}, 0.618)`,
      boxShadow: "1px 1px 1px #6661"
    },
    "&::after": {
      left: "0",
      top: "0",
      position: "absolute",
      display: "inline",
      content: "",
      width: "calc(100% + 4px)",
      height: "calc(100% + 2px) ",
      marginLeft: "-2px",
      marginTop: "-1px",
      borderRadius: "3px",
      backgroundColor: `rgba(${$}, 0.2)`,
      zIndex: -1
    }
  }
}), ht = ({
  searchString: e = "",
  content: t,
  stringLength: n = 150
}) => {
  const a = t.toLowerCase().indexOf(e.toLowerCase());
  if (a !== -1) {
    let o = Math.max(0, a - n / 2), i = Math.min(
      t.length,
      a + e.length + n / 2
    );
    if (o > 0) {
      const s = t.lastIndexOf(" ", o + 1);
      o = s > 0 ? s + 1 : o;
    }
    if (i < t.length) {
      const s = t.indexOf(" ", i - 1);
      i = s > -1 ? s : i;
    }
    t = (o > 0 ? "... " : "") + t.substring(o, i) + (i < t.length ? " ..." : "");
  } else
    t = t.substring(0, n) + "...";
  return t;
}, re = ({
  body: e,
  stringLength: t,
  type: n = "content"
}) => {
  const { state: a } = S(w), { searchString: o } = a;
  let i = String(e.value);
  n == "snippet" && (i = ht({
    searchString: o,
    content: i,
    stringLength: t
  }));
  let s;
  const u = ce(i);
  if (e.format === "text/plain") {
    const g = /\n/g;
    s = i == null ? void 0 : i.replace(g, "<br />");
  }
  if (e.format === "text/markdown" && (s = u == null ? void 0 : u.html), e.format === "text/html" && (s = i), o && s) {
    const g = new RegExp(`(${o})`, "gi");
    s = s == null ? void 0 : s.replace(
      g,
      (h) => `<span class="highlight">${h}</span>`
    );
  }
  const l = [e.id, n].join("-"), p = ["ar"].includes(String(e.language)), m = p ? "rtl" : "ltr", c = p ? "1.3em" : "1em";
  return s ? /* @__PURE__ */ r.createElement(
    pt,
    {
      dangerouslySetInnerHTML: { __html: s },
      "data-body-id": l,
      "data-testid": "scroll-item-body",
      id: l,
      dir: m,
      css: { fontSize: c }
    }
  ) : null;
}, gt = ({
  hasItemBreak: e,
  isLastItem: t,
  item: n,
  itemCount: a,
  itemNumber: o
}) => {
  var c;
  const { state: i } = r.useContext(w), { annotations: s, vault: u } = i, l = u == null ? void 0 : u.get(n), p = (c = s == null ? void 0 : s.filter((g) => g.target === n.id)) == null ? void 0 : c.map((g) => {
    var h;
    return (h = g == null ? void 0 : g.body) == null ? void 0 : h.map((x, E) => /* @__PURE__ */ r.createElement(
      re,
      {
        body: x,
        key: E
      }
    ));
  }), m = {
    current: o,
    total: a
  };
  return /* @__PURE__ */ r.createElement(
    Ae,
    {
      "data-page-break": e,
      "data-page-number": o,
      "data-last-item": t
    },
    /* @__PURE__ */ r.createElement(J, null, l && /* @__PURE__ */ r.createElement(mt, { canvas: l, canvasInfo: m })),
    /* @__PURE__ */ r.createElement(_e, null, (l == null ? void 0 : l.label) && /* @__PURE__ */ r.createElement("strong", null, /* @__PURE__ */ r.createElement(F, { label: l == null ? void 0 : l.label }), " ", `(${m.current} / ${m.total})`), /* @__PURE__ */ r.createElement("div", null, p || /* @__PURE__ */ r.createElement("p", null, "[Blank]")), e && /* @__PURE__ */ r.createElement(Pe, { "aria-label": "Page Break" }))
  );
}, ft = r.memo(gt), z = "255, 197, 32", xt = d("span", {
  fontWeight: "700"
}), vt = d("div", {
  display: "flex",
  gap: "0.25rem"
}), wt = d("div", {
  fontSize: "0.9rem",
  padding: "1rem 0.618rem"
}), yt = d("div", {
  display: "flex",
  flexDirection: "column",
  gap: "0.5rem",
  hr: {
    margin: "0.618rem",
    border: "none",
    borderBottom: "1px solid #6662"
  },
  button: {
    backgroundColor: "#6660",
    opacity: "0.7",
    transition: "$all",
    padding: "0.5rem 0.618rem",
    fontSize: "0.9rem",
    lineHeight: "1.1rem",
    textAlign: "left",
    borderRadius: "6px",
    border: "1px solid #6662",
    display: "flex",
    flexDirection: "column",
    gap: "0.25rem",
    "&:hover": {
      opacity: "1",
      boxShadow: "5px 5px 13px #6662"
    },
    "&[data-result=true]": {
      backgroundColor: `rgba(${z}, 0.2)`,
      borderColor: `rgba(${z}, 0.2)`,
      opacity: "1",
      "&:hover": {
        backgroundColor: `rgba(${z}, 0.2)`
      }
    }
  }
}), ne = d("input", {
  margin: "0",
  padding: "0 1rem 0 2rem",
  background: "none",
  zIndex: "2",
  height: "2rem",
  marginLeft: "1rem",
  marginTop: "1rem",
  justifyContent: "center",
  display: "flex",
  alignItems: "center",
  fontSize: "1rem",
  borderRadius: "2rem",
  fontFamily: "inherit",
  backgroundColor: "$primary",
  border: "none",
  color: "$secondary",
  cursor: "text",
  filter: "drop-shadow(2px 2px 5px #0003)",
  transition: "$all",
  boxSizing: "content-box !important",
  flexGrow: "0",
  width: "100%",
  "&:placeholder": {
    color: "inherit"
  }
}), oe = d("span", {
  position: "absolute",
  zIndex: "3",
  width: "2rem",
  height: "2rem",
  padding: "8px",
  marginTop: "1rem",
  marginLeft: "1rem",
  color: "$secondary",
  fill: "$secondary",
  stroke: "$secondary",
  transition: "$all",
  cursor: "text"
}), St = d("button", {
  opacity: "1",
  display: "flex",
  alignItems: "center",
  borderRadius: "2rem",
  width: "2rem",
  height: "2rem",
  alignSelf: "center",
  marginTop: "1rem",
  gap: "0.35rem",
  backgroundColor: "$accent",
  fill: "$secondary",
  flexShrink: "0",
  svg: {
    padding: "6px",
    color: "inherit",
    fill: "inherit"
  },
  '&[aria-disabled="true"]': {
    opacity: "0",
    display: "none"
  }
}), Et = d("form", {
  display: "flex",
  justifyContent: "space-between",
  gap: "1rem",
  width: "100%",
  variants: {
    isPanelExpanded: {
      true: {
        [`${oe}`]: {
          marginLeft: "0"
        },
        [`& ${ne}`]: {
          marginLeft: "0",
          backgroundColor: "$primary",
          width: "auto",
          flexGrow: "1"
          //
        }
      },
      false: {
        //
      }
    }
  }
}), bt = d("div", {}), N = ({
  annotation: e,
  isResult: t
}) => {
  var o, i;
  const n = [(o = e == null ? void 0 : e.body) == null ? void 0 : o.id, "content"].join("-"), a = (s) => {
    var u;
    s && ((u = document.getElementById(s)) == null || u.scrollIntoView({ behavior: "smooth" }));
  };
  return /* @__PURE__ */ r.createElement(
    "button",
    {
      "data-result": t ? "true" : "false",
      onClick: () => a(n),
      key: n
    },
    /* @__PURE__ */ r.createElement(vt, null, /* @__PURE__ */ r.createElement(xt, null, e.motivation), /* @__PURE__ */ r.createElement("span", null, (i = e == null ? void 0 : e.body) == null ? void 0 : i.language)),
    /* @__PURE__ */ r.createElement(
      re,
      {
        body: e.body,
        stringLength: 144,
        type: "snippet"
      }
    )
  );
}, It = ({
  results: e
}) => /* @__PURE__ */ r.createElement(yt, null, e.found.map((t) => /* @__PURE__ */ r.createElement(
  N,
  {
    annotation: t,
    isResult: !0,
    key: t.id
  }
)), /* @__PURE__ */ r.createElement("hr", null), e.notFound.map((t) => /* @__PURE__ */ r.createElement(
  N,
  {
    annotation: t,
    key: t.id
  }
))), Ct = {
  charset: "latin:extra, arabic:extra, cyrillic:extra, cjk:extra",
  optimize: !0,
  tokenize: "full",
  resolution: 9,
  document: {
    id: "id",
    index: "content"
  }
}, kt = () => {
  var l;
  const { state: e } = S(w), { annotations: t, searchString: n = "" } = e, a = new de.Document(Ct), o = [];
  t == null || t.forEach((p) => {
    var m;
    (m = p == null ? void 0 : p.body) == null || m.forEach((c) => {
      var h;
      const g = (h = c == null ? void 0 : c.value) == null ? void 0 : h.replace(/\n/g, "");
      o.push(c == null ? void 0 : c.id), a.add({
        id: c == null ? void 0 : c.id,
        content: g
      });
    });
  });
  function i(p) {
    return p.map((m) => t == null ? void 0 : t.filter((c) => c.body.find((g) => g.id === m)).map((c) => {
      const g = c.body.findIndex((h) => h.id === m);
      return {
        ...c,
        body: c.body[g]
      };
    }).shift());
  }
  const s = a == null ? void 0 : a.search(n).reduce((p, m) => [.../* @__PURE__ */ new Set([...p, ...m.result])], []), u = {
    found: i(s),
    notFound: i(
      o.filter((p) => !s.includes(p))
    )
  };
  return /* @__PURE__ */ r.createElement(bt, null, n && /* @__PURE__ */ r.createElement(wt, null, (l = u.found) == null ? void 0 : l.length, " results for ", /* @__PURE__ */ r.createElement("strong", null, n)), /* @__PURE__ */ r.createElement(It, { results: u }));
}, $t = () => /* @__PURE__ */ r.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", viewBox: "0 0 512 512" }, /* @__PURE__ */ r.createElement("title", null, "Search"), /* @__PURE__ */ r.createElement("path", { d: "M456.69 421.39L362.6 327.3a173.81 173.81 0 0034.84-104.58C397.44 126.38 319.06 48 222.72 48S48 126.38 48 222.72s78.38 174.72 174.72 174.72A173.81 173.81 0 00327.3 362.6l94.09 94.09a25 25 0 0035.3-35.3zM97.92 222.72a124.8 124.8 0 11124.8 124.8 124.95 124.95 0 01-124.8-124.8z" })), zt = () => /* @__PURE__ */ r.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", viewBox: "0 0 512 512" }, /* @__PURE__ */ r.createElement("title", null, "Close"), /* @__PURE__ */ r.createElement("path", { d: "M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z" })), Lt = ({
  togglePanel: e,
  isPanelExpanded: t
}) => {
  const { dispatch: n, state: a } = S(w), { searchString: o } = a, i = B(null), s = () => i.current === document.activeElement && e(!0), u = () => {
    var h;
    return (h = i == null ? void 0 : i.current) == null ? void 0 : h.focus();
  }, l = () => {
    var h;
    (h = i.current) == null || h.blur(), p(), e(!1), n({
      payload: "",
      type: "updateSearchString"
    });
  }, p = () => {
    i.current && (i.current.value = "");
  }, m = (h) => {
    h.preventDefault(), l();
  }, c = (h) => {
    h.key === "Escape" && (h.preventDefault(), l());
  };
  v(() => (document == null || document.addEventListener("keydown", c), () => {
    document == null || document.removeEventListener("keydown", c);
  }), []), v(() => {
    if (i != null && i.current)
      return i.current.addEventListener("focus", s), i.current.addEventListener("blur", s), () => {
        i.current && (i.current.removeEventListener("focus", s), i.current.removeEventListener("blur", s));
      };
  }, []);
  const g = (h) => {
    var x;
    n({
      payload: (x = h == null ? void 0 : h.target) == null ? void 0 : x.value,
      type: "updateSearchString"
    });
  };
  return /* @__PURE__ */ r.createElement(
    Et,
    {
      id: "scroll-search",
      autoComplete: "off",
      isPanelExpanded: t
    },
    /* @__PURE__ */ r.createElement(oe, { onClick: u }, /* @__PURE__ */ r.createElement($t, null)),
    /* @__PURE__ */ r.createElement(
      ne,
      {
        ref: i,
        name: "clover-search",
        type: "text",
        placeholder: "Search...",
        defaultValue: o,
        onChange: g
      }
    ),
    /* @__PURE__ */ r.createElement(
      St,
      {
        "aria-disabled": !t,
        "aria-label": "Close search panel",
        onClick: m,
        disabled: !t
      },
      /* @__PURE__ */ r.createElement(zt, null)
    )
  );
}, At = d("div", {
  display: "flex",
  flexDirection: "column",
  flexGrow: "1",
  flexShrink: "0",
  position: "relative",
  zIndex: "1",
  maxWidth: "100%",
  marginTop: "1rem",
  transition: "$all",
  variants: {
    isPanelExpanded: {
      true: {
        zIndex: 1,
        opacity: 1,
        transform: "translateX(0)"
      },
      false: {
        zIndex: -1,
        opacity: 0,
        transform: "translateX(-2.618rem)",
        transitionDelay: "0",
        transition: "none"
      }
    }
  }
}), _t = (e) => {
  const [t, n] = I({ top: 0, left: 0 });
  return v(() => {
    const a = () => {
      if (e.current) {
        const o = e.current.getBoundingClientRect();
        n({
          top: o.top,
          left: o.left
        });
      }
    };
    return window.addEventListener("scroll", a), window.addEventListener("resize", a), a(), () => {
      window.removeEventListener("scroll", a), window.removeEventListener("resize", a);
    };
  }, [e]), t;
}, Pt = ({
  isPanelExpanded: e,
  handlePanel: t
}) => {
  var g;
  const n = B(null), { state: a } = S(w), { options: o } = a, { offset: i } = o, { top: s } = _t(n), u = s ? s < i : !1, l = (g = n == null ? void 0 : n.current) == null ? void 0 : g.offsetWidth, p = l && l * 0.5, m = p ? p - 315 : 180, c = {
    top: u ? i : 0,
    width: `calc(${p}px - 1.318rem)`,
    maxWidth: e ? "100%" : `${m}px`,
    minWidth: "130px"
  };
  return /* @__PURE__ */ r.createElement(
    ke,
    {
      ref: n,
      className: u ? "anchor" : "",
      "data-testid": "scroll-panel"
    },
    /* @__PURE__ */ r.createElement(_, { style: c }, /* @__PURE__ */ r.createElement(
      Lt,
      {
        togglePanel: t,
        isPanelExpanded: e
      }
    ), /* @__PURE__ */ r.createElement(
      At,
      {
        "data-testid": "scroll-panel-results",
        "data-panel-expanded": e,
        isPanelExpanded: e
      },
      e && /* @__PURE__ */ r.createElement(kt, null)
    ))
  );
}, Bt = ({ items: e }) => {
  const [t, n] = I(!1), a = (o) => n(o);
  return /* @__PURE__ */ r.createElement(r.Fragment, null, /* @__PURE__ */ r.createElement(
    Pt,
    {
      isPanelExpanded: t,
      handlePanel: a
    }
  ), /* @__PURE__ */ r.createElement(Be, { "data-figures-visible": !t }, e.map((o, i) => {
    const s = i + 1, u = s === e.length;
    return /* @__PURE__ */ r.createElement(
      ft,
      {
        item: o,
        hasItemBreak: s < e.length,
        isLastItem: u,
        key: o.id,
        itemCount: e.length,
        itemNumber: s
      }
    );
  })));
}, Mt = (e, t) => {
  const [n, a] = I([]);
  return v(() => {
    if (!t)
      return;
    const o = [];
    e == null || e.forEach((s) => {
      var l;
      const u = t.get(s);
      (l = u == null ? void 0 : u.annotations) == null || l.forEach(
        (p) => {
          var c;
          const m = t.get(p);
          (c = m == null ? void 0 : m.items) == null || c.forEach(
            (g) => {
              var x;
              const h = t.get(g);
              h && o.push({
                ...h,
                body: (x = h == null ? void 0 : h.body) == null ? void 0 : x.map((E) => t.get(E))
              });
            }
          );
        }
      );
    });
    const i = o.reduce(
      (s, u) => (s.some((l) => l.id === u.id) || s.push(u), s),
      []
    );
    a(i);
  }, [e, t]), n;
}, Tt = ({ iiifContent: e }) => {
  const [t, n] = I(), { state: a, dispatch: o } = S(w), { vault: i } = a, s = Mt(t == null ? void 0 : t.items, i);
  return v(() => {
    i && i.load(e).then((u) => u && n(u)).catch(
      (u) => console.error(`Manifest ${e} failed to load: ${u}`)
    );
  }, [e, i]), v(() => {
    o({
      type: "updateAnnotations",
      payload: s
    });
  }, [s, o]), t ? /* @__PURE__ */ r.createElement(ze, null, t.label && /* @__PURE__ */ r.createElement(Le, { label: t.label }), t.items && /* @__PURE__ */ r.createElement(Bt, { items: t.items })) : null;
}, Kt = ({
  iiifContent: e,
  options: t
}) => /* @__PURE__ */ r.createElement(me, { options: t }, /* @__PURE__ */ r.createElement(Tt, { iiifContent: e }));
export {
  Kt as default
};
